src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs

```diff
@@ -130,79 +141,78 @@ internal static void HtmlEncode(string? value, TextWriter output)
-
-                    if (count > 0)
-                    {
-                        b.Append(value, startIndex, count);
-                    }
-
-                    startIndex = i + 1;
-                    count = 0;
+                return addDoubleQuotes ? "\"" + value + "\"" : value;
+            }
 
-                    switch (c)
-                    {
-                        case '\r':
-                            b.Append("\\r");
-                            break;
-                        case '\t':
-                            b.Append("\\t");
-                            break;
-                        case '\"':
-                            b.Append("\\\"");
-                            break;
-                        case '\\':
-                            b.Append("\\\\");
-                            break;
-                        case '\n':
-                            b.Append("\\n");
-                            break;
-                        case '\b':
-                            b.Append("\\b");
-                            break;
-                        case '\f':
-                            b.Append("\\f");
-                            break;
-                        default:
-                            AppendCharAsUnicodeJavaScript(b, c);
-                            break;
-                    }
-                }
-                else
+            StringBuilder sb;
+            if (addDoubleQuotes)
+            {
+                sb = new StringBuilder(value.Length + 7);
+                sb.Append('"');
+            }
+            else
+            {
+                sb = new StringBuilder(value.Length + 5);
+            }
+            int startIndex = 0;
+            do
+            {
+                sb.Append(value, startIndex, i);
+                char c = value[startIndex + i];
+                switch (c)
                 {
-                    count++;
+                    case '\r':
+                        sb.Append("\\r");
+                        break;
+                    case '\t':
+                        sb.Append("\\t");
+                        break;
+                    case '\"':
+                        sb.Append("\\\"");
+                        break;
+                    case '\\':
+                        sb.Append("\\\\");
+                        break;
+                    case '\n':
+                        sb.Append("\\n");
+                        break;
+                    case '\b':
+                        sb.Append("\\b");
+                        break;
+                    case '\f':
+                        sb.Append("\\f");
+                        break;
+                    default:
+                        AppendCharAsUnicodeJavaScript(sb, c);
+                        break;
                 }
-            }
 
-            if (b == null)
+                startIndex += i + 1;
+                i = value.AsSpan(startIndex).IndexOfAny(s_invalid_JavaScriptChars);
```

Comment:
Use `value` as span in the loop too, so that relative indices (only `i` instead of `startIndex + i`) can be used?
Thus only slicing once, then re-using that slice instead of array + slice together.